apps/backend/src/action/logic/check-tool-has-broken.ts

```diff
@@ -1,27 +1,133 @@
+import type { EventSource } from "~/models/event";
+import { PlayerToolState, Tool, ToolState } from "~/models/tool";
+import { always, assoc, error } from "~/utils";
+
+export const InitialPlayerToolState: PlayerToolState = {
+  [Tool.Cart]: ToolState.NotBroken,
+  [Tool.Lamp]: ToolState.NotBroken,
+  [Tool.Pickaxe]: ToolState.NotBroken,
+};
+
+const EventSourceReadError = error("EventSourceReadError");
+const ToolHasBeenBrokenError = error("ToolHasBeenBrokenError");
+const ToolHasNotBeenBrokenError = error("ToolHasNotBeenBrokenError");
+
+type EventSourceReadError = ReturnType<typeof EventSourceReadError>;
+type ToolHasBeenBrokenError = ReturnType<typeof ToolHasBeenBrokenError>;
+type ToolHasNotBeenBrokenError = ReturnType<typeof ToolHasNotBeenBrokenError>;
+
+export type CheckToolHasBrokenError =
+  | EventSourceReadError
+  | ToolHasBeenBrokenError
+  | ToolHasNotBeenBrokenError;
+
+type PlayerToolStateOrError = E.Either<
+  CheckToolHasBrokenError,
+  PlayerToolState
+>;
 
 export interface CheckToolHasBroken {
   (repository: EventSource<Event>, command: BrokenToolCommand): ResultAsync<
     boolean,
-    Error
+    CheckToolHasBrokenError
   >;
 }
 
+const readAllEventsFromEventSource = (source: EventSource<Event>) =>
+  ResultAsync.fromPromise(
+    source.read(),
+    always(EventSourceReadError("failed to read events from source"))
+  );
+
+const filterAllEventsByPlayerId =
+  (command: BrokenToolCommand) => (event: Event) =>
+    match(event.data)
+      .with({ playerId: command.data.playerId }, always(true))
+      .otherwise(always(false));
+
+const validatePlayerToolCanBeBroken =
+  (stateOrError: PlayerToolStateOrError) =>
+  ({ data: { playerId, tool } }: BrokenToolHasBeenPlacedEvent) =>
+    pipe(
+      stateOrError,
+      E.chain((state) =>
+        state[tool] === ToolState.NotBroken
+          ? E.right(assoc(tool, ToolState.Broken, state))
+          : E.left(
+              ToolHasBeenBrokenError(
+                `can not broke player ${playerId} tool ${tool}`
+              )
+            )
+      )
+    );
+
+const validatePlayerToolCanBeFixed =
+  (stateOrError: PlayerToolStateOrError) =>
+  ({ data: { playerId, tool } }: BrokenToolHasBeenRemovedEvent) =>
+    pipe(
+      stateOrError,
+      E.chain((state) =>
+        state[tool] === ToolState.Broken
+          ? E.right(assoc(tool, ToolState.NotBroken, state))
+          : E.left(
+              ToolHasNotBeenBrokenError(
+                `can not fix player ${playerId} tool ${tool}`
+              )
+            )
+      )
+    );
+
+const updatePlayerToolState = (
+  stateOrError: PlayerToolStateOrError,
+  event: Event
+) =>
+  match(event)
+    .when(
+      isBrokenToolHasBeenPlacedEvent,
+      validatePlayerToolCanBeBroken(stateOrError)
+    )
+    .when(
+      isBrokenToolHasBeenRemovedEvent,
+      validatePlayerToolCanBeFixed(stateOrError)
+    )
+    .otherwise(always(stateOrError));
+
+const aggregateAllEventsToGetToolsHasBeenBrokenOnPlayer = (
```

Comment:
Output of this function should return `Tool[]`

Reply:
That's a great approach,
but encountering unexpected consecutive removal or placement events can lead to inaccurate state,
although the probability of such an occurrence is so low that it can be ignored.